pkg/radix/tree.go

```diff
@@ -61,79 +70,55 @@ func (n *node) replaceEdge(e edge) {
 }
 
 func (n *node) getEdge(label byte) *node {
-	num := len(n.edges)
-	// Use linear search for smaller arrays
-	if num < 64 {
-		for i := 0; i < num; i++ {
-			if n.edges[i].label == label {
-				return n.edges[i].node
+	// linear search for small slices
+	if len(n.edges) < 16 {
+		for _, e := range n.edges {
+			if e.label == label {
+				return e.node
 			}
 		}
-
 		return nil
 	}
 
-	idx := sort.Search(num, func(i int) bool {
-		return n.edges[i].label >= label
-	})
-	if idx < num && n.edges[idx].label == label {
-		return n.edges[idx].node
-	}
-	return nil
-}
-
-func (n *node) delEdge(label byte) {
+	// binary search for larger
 	num := len(n.edges)
-	idx := sort.Search(num, func(i int) bool {
-		return n.edges[i].label >= label
-	})
-	if idx < num && n.edges[idx].label == label {
-		copy(n.edges[idx:], n.edges[idx+1:])
-		n.edges[len(n.edges)-1] = edge{}
-		n.edges = n.edges[:len(n.edges)-1]
+	i, j := 0, num
+	for i < j {
+		h := int(uint(i+j) >> 1)
+		if n.edges[h].label < label {
+			i = h + 1
+		} else {
+			j = h
+		}
+	}
+	if i < num && n.edges[i].label == label {
+		return n.edges[i].node
 	}
+	return nil
 }
 
 type edges []edge
 
-func (e edges) Len() int {
-	return len(e)
-}
-
-func (e edges) Less(i, j int) bool {
-	return e[i].label < e[j].label
-}
-
-func (e edges) Swap(i, j int) {
-	e[i], e[j] = e[j], e[i]
-}
-
-func (e edges) Sort() {
-	sort.Sort(e)
-}
-
 // Tree implements a radix tree. This can be treated as a
 // Dictionary abstract data type. The main advantage over
 // a standard hash map is prefix-based lookups and
-// ordered iteration.  The tree is safe for concurrent access.
```

Comment:
I'd rather we keep this goroutine safe.  The existing code that uses this relies on this type being safe for concurrent access.  Otherwise, we'll have to perform these operations under a separate mutex or one that already exists in `Engine`.  The spots in the `Engine` where this would be called only need an `RLock` and this would need a `Lock`.  If it's a separate mutex, we open up the possibility of deadlocks.

Reply:
Oops, yep. I wasn't careful enough at validating the engine side locks.